Appending target framework to the OutputPath/IntermediateOutputPath

adalon opened the pull request:
We only can do this when AndroidUseLatestPlatformSdk is false due to the
need of evaluating the MsBuild properties before calculating the latest SDK
using the _SetLatestTargetFrameworkVersion target (Xamarin.Android.Common.targets)

XA heavily leverages generated files in the intermediate output directory to achieve incremental build,
which is key since a lot of the Android/Java SDK tools are slow.

In XA, targeting different API levels is done by changing the TargetFramework property in the project.
This effectively invalidates all previous intermediate artifacts. Since this is not an uncommon
action (i.e. user is trying out various target API levels to see which one works best for their needs),
we'd like to preserve whatever artifacts were generated in a previous run for a given framework.

So rather than invalidating all incremental files when that property is changed (i.e. doing a full
clean+build on the next run), a better approach is to append the target framework to the intermediate output path,
following in the footsteps from .NETCore/.NETStandard (which always append the target framework too).

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/xamarin/xamarin-android?pullRequest=1628) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/xamarin/xamarin-android?pullRequest=1628) before we can accept your contribution.<br/><br/>:x: adalon [sign now](https://cla.dotnetfoundation.org/xamarin/xamarin-android?pullRequest=1628)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/xamarin/xamarin-android?pullRequest=1628) it.</sub>

jonpryor commented on src/Xamarin.Android.Build.Tasks/Xamarin.Android.TargetFrameworkInference.targets:
```diff
@@ -0,0 +1,19 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Please place this file into [`src/Xamarin.Android.Build.Tasks/MSBuild/Xamarin/Android`](https://github.com/xamarin/xamarin-android/tree/master/src/Xamarin.Android.Build.Tasks/MSBuild/Xamarin/Android).

adalon replied:
ok, I can assume that the file will be deployed/installed in the same directory where the Xamarin.Android.CSharp.targets is located, right? @dellis1972 @jonpryor

jonpryor replied:
Yup, so long as as the [correct `%(Link)` metadata is provided](https://github.com/xamarin/xamarin-android/blob/master/src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.targets#L45-L49).

jonpryor commented on src/Xamarin.Android.Build.Tasks/Xamarin.Android.TargetFrameworkInference.targets:
```diff
@@ -0,0 +1,19 @@
+	</PropertyGroup>
+	<PropertyGroup Condition="'$(AppendTargetFrameworkToOutputPath)' == 'true'">
+		<OutputPath Condition="'$(OutputPath)' != '' and !HasTrailingSlash('$(OutputPath)')">$(OutputPath)\</OutputPath>
+		<OutputPath>$(OutputPath)$(TargetFrameworkToOutputPath)</OutputPath>
```
I can understand wanting to update `$(IntermediateOutputPath)`, but *`$(OutputPath)`*?!

I find it hard to believe that changing `$(OutputPath)` won't break large swaths of the world, and as `$(AppendTargetFrameworkToOutputPath)` is default True, that's quite a risk.

adalon replied:
I think we can avoid the OutputPath for now since the IntermediateOutputPath should be the one that is causing the 99% of the issues with the generated artifacts.

I will remove it and update the PR.

jonpryor commented:
[The build is broken](https://jenkins.mono-project.com/job/xamarin-android-pr-builder/3037/):

```
The "CreateNativeLibraryArchive" task failed unexpectedly. [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/Mono.Posix/Mono.Posix.csproj]
System.IO.FileNotFoundException: /Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/bin/Debug/lib/xamarin.android/xbuild-frameworks/MonoAndroid/v1.0/MonoAndroid81/../../../xbuild/Xamarin/Android/lib/armeabi-v7a/libMonoPosixHelper.so does not exist [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/Mono.Posix/Mono.Posix.csproj]
File name: '/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/bin/Debug/lib/xamarin.android/xbuild-frameworks/MonoAndroid/v1.0/MonoAndroid81/../../../xbuild/Xamarin/Android/lib/armeabi-v7a/libMonoPosixHelper.so' [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/Mono.Posix/Mono.Posix.csproj]
  at System.IO.File.Copy (System.String sourceFileName, System.String destFileName, System.Boolean overwrite) [0x00193] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.IO/File.cs:111  [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/Mono.Posix/Mono.Posix.csproj]
  at Xamarin.Android.Tools.Files.CopyIfChanged (System.String source, System.String destination) [0x00036] in /Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/Xamarin.Android.Build.Tasks/Utilities/Files.cs:63  [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/Mono.Posix/Mono.Posix.csproj]
  at Xamarin.Android.Tasks.MonoAndroidHelper.CopyIfChanged (System.String source, System.String destination) [0x00001] in /Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/Xamarin.Android.Build.Tasks/Utilities/MonoAndroidHelper.cs:386  [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/Mono.Posix/Mono.Posix.csproj]
  at Xamarin.Android.Tasks.CreateNativeLibraryArchive.Execute () [0x000f7] in /Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/Xamarin.Android.Build.Tasks/Tasks/CreateNativeLibraryArchive.cs:62  [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/Mono.Posix/Mono.Posix.csproj]
  at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute () [0x00023] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/msbuild-15/src/Build/BackEnd/TaskExecutionHost/TaskExecutionHost.cs:631  [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/Mono.Posix/Mono.Posix.csproj]
  at Microsoft.Build.BackEnd.TaskBuilder+<ExecuteInstantiatedTask>d__26.MoveNext () [0x00212] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/msbuild-15/src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs:787  [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/Mono.Posix/Mono.Posix.csproj]
```

Most likely cause is because `$(OutputPath)` was updated; this path:

```
…/bin/Debug/lib/xamarin.android/xbuild-frameworks/MonoAndroid/v1.0/MonoAndroid81/../../../xbuild/Xamarin/Android/lib/armeabi-v7a/libMonoPosixHelper.so
```

should be:

```
…/bin/Debug/lib/xamarin.android/xbuild-frameworks/MonoAndroid/v1.0/../../../xbuild/Xamarin/Android/lib/armeabi-v7a/libMonoPosixHelper.so
```

and this breaks because the [`@(EmbeddedNativeLibrary)` declaration uses `$(OutputPath)`-relative paths](https://github.com/xamarin/xamarin-android/blob/b0efc6ad8a79d592c2771628feba1a21a036d6cd/src/Mono.Posix/Mono.Posix.csproj#L259-L278).

*At minimum*, `Mono.Posix.csproj` will need export `$(AppendTargetFrameworkToOutputPath)`=False, but -- [as pondered earlier](#pullrequestreview-117048353) -- I think this is a sign that changing `$(OutputPath)` will have *lots* of unanticipated consequences, and perhaps should be reconsidered.

adalon commented:
@jonpryor @dellis1972 @kzu I just updated the PR with the requested changes by @jonpryor 

- Moved Xamarin.Android.TargetFrameworkInference.targets to Android.Build.Tasks/MSBuild/Xamarin/Android directory
- And avoid changing the OutputPath property by default

jonpryor commented:
@dellis1972: any idea why there's a unit test failure? https://jenkins.mono-project.com/job/xamarin-android-pr-builder/3074/testReport/junit/Xamarin.Android.Build.Tests.BuildTest/BuildMultiDexApplication/Xamarin_Android_Build_Tests_BuildTest_BuildMultiDexApplication_False__v7_1_____Debug/

dellis1972 commented:
@adalon I think there are a few build issues on the bots. not 100% sure. 

That build failure is probably a unit test issue. It will be looking for files in `obj/Debug/android/bin` but I suspect that folder path has changed due to this PR. 

I was expecting a ton of tests to fail to be honest..

adalon commented:
@dellis1972 we really need this change to be merged asap. Could you please review the asserts we're doing in the tests and fixed them to look into the obj\Debug\{TF}\* location instead?

dellis1972 commented:
I'm gonna investigate the unit test failures today

adalon commented:
@dean we're ready to go!

jonpryor commented:
This is moderately disturbing to me: the `<Import/>` for `Xamarin.Android.TargetFrameworkInference.targets` occurs *before* the `<Import/>` for `Microsoft.CSharp.targets`, meaning at import time `$(BaseIntermediateOutputPath)` *will **never** be set* when `Xamarin.Android.TargetFrameworkInference.targets` is executed.

That seems...horrifying?

(Then again, the prospect of `Microsoft.Common.targets` changing the definition of `$(BaseIntermediateOutputPath)` is *equally* horrifying.)

This also doesn't touch `Xamarin.Android.FSharp.targets`. Should it?

It feels like instead of updating `Xamarin.Android.CSharp.targets`, `Xamarin.Android.TargetFrameworkInference.targets` should instead be `<Import/>`ed from `Xamarin.Android.Common.targets`, which means that `$(BaseIntermediateOutputPath)`/etc. *will* be set via `Microsoft.Common.targets` before `Xamarin.Android.TargetFrameworkInference.targets` executes, *and* means `Xamarin.Android.FSharp.targets` doesn't need updating.

kzu commented:
Let's call this new targets `Xamarin.Android.DefaultOutputPaths.targets` instead, otherwise it sounds scarier than it is.

@adalon we're missing [ensuring a trailing slash](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.DefaultOutputPaths.targets#L40) on the `BaseIntermediateOutputPath`, so after the first property group, we should have:

```xml
<PropertyGroup>
   <BaseIntermediateOutputPath Condition="!HasTrailingSlash('$(BaseIntermediateOutputPath)')">$(BaseIntermediateOutputPath)\</BaseIntermediateOutputPath>
</PropertyGroup>
```

@jonpryor this target essentially does the same as [Microsoft.NET.DefaultOutputPaths.targets](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.DefaultOutputPaths.targets), so, the scary part shouldn't be scary at all, since it's exactly how it's evaluated for C# projects in SDK-style projects (that is, evaluated *before* `Common.targets` by being a [BeforeCommon.targets](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets#L27) imported [before the language targets](https://github.com/dotnet/sdk/blob/e9ebf32ca6ab22f56c3b50e9dd4c6ccfcc19d220/src/Tasks/Microsoft.NET.Build.Tasks/sdk/Sdk.targets#L25-L26)).

Essentially, we're providing an improved default value *only* if the user hasn't overriden them in his project, so that by the time the language targets [start evaluating path-related properties](https://github.com/Microsoft/msbuild/blob/master/src/Tasks/Microsoft.Common.CurrentVersion.targets#L348) our default is used throughout. So it is very much by design that we want this to evaluate *before* that.

adalon commented:
@kzu done

jonathanpeppers commented:
@jonpryor we can ignore the Windows build failure, because NUnit is hanging...

The MSBuild tests passed:
```
Test Run Summary
    Overall result: Warning
    Test Count: 314, Passed: 275, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 39
      Skipped Tests - Ignored: 39, Explicit: 0, Other: 0
    Start time: 2018-05-14 20:00:28Z
      End time: 2018-05-14 20:23:15Z
      Duration: 1367.518 second
```
NUnit is hanging on master, so I'll look into that separately.

onovotny commented:
This change is causing issues when used with SDK style projects. SDK style projects already append the TFM to the intermediate and output paths. This new behavior causes `monoandroid70\MonoAndroid70` to be the output.

`AppendTargetFrameworkToOutputPath` and `AppendTargetFrameworkToIntermediateOutputPath` is set to `true` here:

https://github.com/dotnet/sdk/blob/33d51eab4d1e8dccf90eec946cf311ef29e0857f/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.TargetFrameworkInference.targets#L177-L180

At the very least, the Xamarin targets should have an extra condition to know if they're being used in an SDK style project or not. As it's using the same MSBuild property, there's no way to workaround this.

kzu commented:
@onovotny isn't the workaround simply to set `AppendTargetFrameworkToOutputPath=false` before importing the Xamarin targets, from the Sdk.targets in the MSBuildExtras.Sdk?

Agreed that we should turn off this for SDK-style. Will send another PR.

onovotny commented:
No, because that'll affect the SDK's use of that property too and it'll dump stuff in to `\bin\Debug` instead of `bin\debug\monoandroid80`.

kzu commented:
You'd use the new `$(EnableDefaultOutputPaths)` to turn this behavior off, from https://github.com/xamarin/xamarin-android/commit/1589dff3c9e38f0761f65c4e0bbe33a21acd8589. Would that be enough?

kzu commented:
Yes.

jonathanpeppers commented:
@kzu I tried using `$(AppendTargetFrameworkToOutputPath)` and it does not work on master, this `PropertyGroup` doesn't check for it: https://github.com/xamarin/xamarin-android/blob/ddb647542488cfd1469923c32d87bf60fbb6eca3/src/Xamarin.Android.Build.Tasks/MSBuild/Xamarin/Android/Xamarin.Android.DefaultOutputPaths.targets#L21-L24

This change has caused a some grief (other than what @onovotny mentions):
- VS for Mac had issues - we had to disable on Mac
- We are using 13 extra characters of our `MAX_PATH` budget, this breaks this test project on my machine: https://github.com/xamarin/xamarin-android/tree/master/tests/Xamarin.Forms-Performance-Integration
- There are odd paths such as `obj\\Debug\MonoAndroid81` with two slashes
- We have to turn this feature off in all of our tests, because they validate things in `obj`... Seems a bit scary to me

Do you think we can look into another approach? Maybe something like #1878 that does a partial clean when `TargetFrameworkVersion` changes?

kzu commented:
@jonathanpeppers in my view, there should be nothing that breaks anywhere when changes are made to those paths, since, by definition, *anyone* can change them, users too. So sure, if tests are hardcoded and need to disable it, we should have an easy mechanism to opt-out, but I see no reason why this shouldn't work as intended, since it's the same mechanism used in multi-targeting SDK-style to separate the obj artifacts for each target framework. I don't see a reason why it couldn't also "just work" for us here :(.

Also, does it work if you use `$(EnableDefaultOutputPaths)` instead?

jonathanpeppers commented:
`obj` paths are only hardcoded in tests, to assert files exists, etc. Perhaps that could have been done in a better way, but this change doesn't update the tests to address it either.

Since [Xamarin.Android.DefaultOutputPaths.targets](https://github.com/xamarin/xamarin-android/blob/ddb647542488cfd1469923c32d87bf60fbb6eca3/src/Xamarin.Android.Build.Tasks/MSBuild/Xamarin/Android/Xamarin.Android.DefaultOutputPaths.targets) is imported last, it needs to be handling any of these properties for them to work:
- `$(AppendTargetFrameworkToOutputPath)`
- `$(EnableDefaultOutputPaths)`

That is why I feel like this change is opening a whole can of worms... Maybe we can do something that is a bit easier and achieve the same result?

I assume the original scenario is something like:
- Build
- Build with a different `TargetFrameworkVersion`
- Something breaks?

There is not an accompanying test showing a concrete example in this PR, so just guessing?

kzu commented:
The goal of the PR is to provide exactly the same behavior as SDK-style per-TF intermediate output paths ;).

The two properties you mention are intended for different things:
* `$(EnableDefaultOutputPaths)`: turn on/off the defaulting wholesale. 
* `$(AppendTargetFrameworkToOutputPath)`: whether to append to the output path, not the intermediate one. It's intended to work exactly as its counterpart in the [SDK](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.TargetFrameworkInference.targets#L170).

This is pretty much the say the SDK targets do their own defaulting too.

jonathanpeppers commented:
> The goal of the PR is to provide exactly the same behavior as SDK-style per-TF intermediate output paths

I'm mostly asking _why_? It seems this changing is breaking more things than it fixes. I am trying to find out what it actually fixes.

### Side note

One of the properties is not quite working correctly as of d15-8/master:

`$(AppendTargetFrameworkToOutputPath)=False` should _also_ trigger `$(IntermediateOutputPath)` to be the same as `$(AppendTargetFrameworkToIntermediateOutputPath)=False`, and it's not right now.

`EnableDefaultOutputPaths=False` is working correctly, I believe.

We should have tests checking these, right?

kzu commented:
Feel free to revert and provide a better approach. This was one thing we
found that fixed a lot of manually found errors that may (or may not) have
been fixed by the other stuff that was changed since then. It was a quick
and effective way in our manual testing to fix those.

Back when we did this, the tests weren't even running or building from VS
and a lot of improvements have happened in that area too, I believe, so it
might be easier to contribute now with proper tests.

On Wed, Jul 11, 2018, 6:16 PM Jonathan Peppers <notifications@github.com>
wrote:

> The goal of the PR is to provide exactly the same behavior as SDK-style
> per-TF intermediate output paths
>
> I'm mostly asking *why*? It seems this changing is breaking more things
> than it fixes. I am trying to find out what it actually fixes.
> Side note
>
> One of the properties is not quite working correctly as of d15-8/master:
>
> $(AppendTargetFrameworkToOutputPath)=False should *also* trigger
> $(IntermediateOutputPath) to be the same as
> $(AppendTargetFrameworkToIntermediateOutputPath)=False, and it's not
> right now.
>
> EnableDefaultOutputPaths=False is working correctly, I believe.
>
> We should have tests checking these, right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/xamarin/xamarin-android/pull/1628#issuecomment-404311987>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKW6xQzJHoBBUHCWibq8Kzd4xBnPcpYks5uFmscgaJpZM4TvqhE>
> .
>
-- 

--
/kzu from mobile

kobbe commented:
When that path is obj/release/monoandroid80 insted of obj/release/. How does one use a post script that uses the .apk file without hard coding the api version into the post script command path? For example moving the api to a different location.

kzu commented:
you can invoke that script from msbuild and pass down the
`$(TargetFrameworkIdentifier)` and version, for example, or make it a
post-build target that copies to a stable location, etc.

On Mon, Sep 3, 2018 at 6:02 AM Hans <notifications@github.com> wrote:

> When that path is obj/release/monoandroid80 insted of obj/release/. How
> does one use a post script that uses the .apk file without hard coding the
> api version into the post script command path? For example moving the api
> to a different location.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/xamarin/xamarin-android/pull/1628#issuecomment-418049659>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKW6xopL7E6Y7gdhSTiJ2qaTwfEeXyfks5uXPAOgaJpZM4TvqhE>
> .
>
-- 

--
/kzu from mobile

jonathanpeppers commented:
@kobbe can you not just use `$(IntermediateOutputPath)`?

Is your script outside of MSBuild?
